File: src/main/java/com/acerat/solidtest/checkout/CheckoutHandler.java
package com.acerat.solidtest.checkout;

import com.acerat.solidtest.checkout.state.*;
import com.acerat.solidtest.customers.Customer;
import com.acerat.solidtest.shoppingcart.Order;

public class CheckoutHandler {
    private CheckoutHandlerUtils checkoutHandlerUtils = new CheckoutHandlerUtils();

    public CheckoutState checkout(CheckoutState checkoutState) {

        Order order = checkoutState.getOrder();
        Customer customer = checkoutHandlerUtils.getCustomer(order);

        //shipping Address
        if (checkoutHandlerUtils.isValidShippingInformation(checkoutState, customer)){
            //TODO isValidShipping Information is True
            return checkoutState;
        }else {
            //TODO do something when isShipping Address is False
            return getCheckoutState();
        }


        //TODO Add logic
        if (checkoutState.getReverseItemsRelease(order)) return checkoutState;

        if (checkoutHandlerUtils.isChechoutState(checkoutState, order, customer)) return checkoutState;
        // Send reserved items
        if (checkoutState.sendReservedItems(order)) return checkoutState;
        return checkoutState;
    }

    private CheckoutState getCheckoutState() {
        return new CheckoutState(null);
    }


}

File: src/main/java/com/acerat/solidtest/checkout/CheckoutHandler.kt
package com.acerat.solidtest.checkout

import com.acerat.solidtest.checkout.state.*
import com.acerat.solidtest.configuration.ApplicationConfiguration
import com.acerat.solidtest.customers.Customer
import com.acerat.solidtest.customers.CustomerRepository
import com.acerat.solidtest.shoppingcart.Order

class CheckoutHandler {
    private val checkoutHandlerUtils = CheckoutHandlerUtils()

    fun checkout(checkoutState: CheckoutState): CheckoutState {

        val order = checkoutState.order
        val customer = getCustomer(order)

        if (checkoutHandlerUtils.isValidShippingInformation(checkoutState, customer!!)) return checkoutState

        //TODO Add logic
        if (checkoutState.getReverseItemsRelease(order)) return checkoutState

        if (checkoutHandlerUtils.isCheckoutState(checkoutState, order, customer)) return checkoutState
        // Send reserved items
        return if (checkoutState.sendReservedItems(order)) checkoutState else checkoutState
    }

    private fun noCheckoutState(): CheckoutState {
        return CheckoutState(null)
    }

    fun getCustomer(order: Order): Customer? {
        // Get customer
        return CUSTOMER_REPOSITORY.get(order.customerId)
    }

    companion object {
        private val CUSTOMER_REPOSITORY = CustomerRepository(ApplicationConfiguration.getConnectionString())
    }


}
